packages/type-utils/src/getAwaitedType.ts

```diff
@@ -0,0 +1,19 @@
+import * as tsutils from 'ts-api-utils';
+import type * as ts from 'typescript';
+
+export function getAwaitedType(
+  checker: ts.TypeChecker,
+  type: ts.Type,
+  node: ts.Node,
+): ts.Type {
+  if (
+    tsutils.isThenableType(checker, node, type) &&
+    tsutils.isTypeReference(type)
+  ) {
+    const awaitedType = type.typeArguments?.[0];
```

Comment:
[needs discussion] This function will always return first type argument of type reference, so it won't catch the following case:

```ts
interface alias<T> extends Promise<any> {foo: 'bar'}

declare const value: alias<number>;
async function foo(): Promise<number> {
  const awaited = await value
  //    ^? const awaited: any
  return value; // <- value should be reported because its awaited version above is of type 'any'
}
```

[playground](https://deploy-preview-8693--typescript-eslint.netlify.app/play/#ts=5.5.2&fileType=.ts&code=JYOwLgpgTgZghgYwgAjgG2HAzgHgCoB8yEAHpCACZbIAKUA9gLbBYQ5wgCeRA3jPfQBcyAOQAjOFBEBfAFCyKEBGkkoE9EFjDIAbugCuEYeky4Q%2BxmOgEA3LOycQCZDH1OwwDS4EAKAJTCdEwsbOaW1sg8ssjI6pracADucMCQFMgAvKjJqboGENHIAPRFMTEAegD8hVAQYPpQIHlohnbSQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tieQEMAZolrRE%2BWNCbooiaNA7RI4MAF8QKoA&tsconfig=&tokens=false)

---

But reinventing this wheel on typescript-eslint side seems a bit scary for me. What do you think about using `checker.getAwaitedType()` here? It is an internal TS function, but that's exactly what we want here. `type-utils` already relies on internal APIs, though I'm not sure we want to rely on them even more...
https://github.com/typescript-eslint/typescript-eslint/blob/bebbd0c274e5009bf6b03b3a382782a3a708248e/packages/type-utils/typings/typescript.d.ts#L3-L20

Reply:
Hi @auvred 
I have confirmed that getAwaitedType exists as an internal API in ts v4.7.4, so it's probably possible.
But I'm also not sure it's a good idea to rely on the internal API.
BTW,  I've personally seen very little code that extends promises, so I wonder if this is an edge case.

Reply:
> `getAwaitedType` is on its way to being exposed publicly in TS

Is there any tracking issue in TS repo? I'd like to subscribe to it if there is :smile: 

> See also [#8693 (comment)](https://github.com/typescript-eslint/typescript-eslint/pull/8693#discussion_r1634143467)

Oops, didn't notice this thread

---

Then I don't mind leaving the current `getAwaitedType` implementation as is, but maybe we should say in some `// TODO` that we should reuse `getAwaitedType` from TS when it becomes part of the public API?

Reply:
> > `getAwaitedType` is on its way to being exposed publicly in TS
> 
> Is there any tracking issue in TS repo? I'd like to subscribe to it if there is 😄

Couldn't find one, but Jake Bailey has mentioned that it's an intention of his several times in conversations... see, for example,  https://discord.com/channels/1026804805894672454/1240360626832937011. Might be worth asking him if there's an issue or something to track.

> maybe we should say in some `// TODO` that we should reuse `getAwaitedType` from TS when it becomes part of the public API?

+1

Reply:
Tip: it's great for your street cred as a developer to file an issue on TypeScript... if any enterprising individual subscribed to this thread wants to [search the TypeScript issue tracker for `getAwaitedType`](https://github.com/Microsoft/TypeScript/issues?q=is%3Aissue+getAwaitedType+) and maybe file an issue if one doesn't exist...

Reply:
I filed an issue on TypeScript.: https://github.com/microsoft/TypeScript/issues/59256

Reply:
> I filed an issue on TypeScript.: [microsoft/TypeScript#59256](https://github.com/microsoft/TypeScript/issues/59256)

Thanks for doing that! (And you already have a PR up too, 😀)

***

Regardless, I still think for now we should proceed with the heuristic-based function in this PR.

1. the newly exposed `getAwaitedType` won't become visible to us until we update to the next version of TS anyway 
1. Even once `getAwaitedType` is exposed, we would need to ensure that, despite not publicly declared, it still works at runtime in older versions of TS we support. 
1. Related previously-internal APIs like `isTypeAssignableTo` _are_ already available to us, but I recall it being said somewhere that we wouldn't introduce them until v8 (presumably for similar TS support reasons)

Reply:
Btw found the reference for 3. https://github.com/typescript-eslint/typescript-eslint/discussions/7936

Reply:
IMO we can use the newly exposed `getAwaitedType` already as long as a maintainer verifies the PR using it works in an older TypeScript version. The method's status in the `.d.ts` files doesn't change whether it'll work in older TS versions.

Reply:
@kirkwaiblinger @JoshuaKGoldberg @auvred 
> IMO we can use the newly exposed getAwaitedType already as long as a maintainer verifies the PR using it works in an older TypeScript version. The method's status in the .d.ts files doesn't change whether it'll work in older TS versions.

Agree! I changed to use the TypeChecker's getAwaitedType. 0d0a86a